src/core/lib/event_engine/workqueue.cc

```diff
@@ -0,0 +1,208 @@
+    : closure_(SelfDeletingClosure::Create(std::move(callback))),
+      enqueued_(
+          grpc_core::ExecCtx::Get()->Now().milliseconds_after_process_epoch()) {
+}
+
+WorkQueue::Storage::Storage(Storage&& other) noexcept
+    : closure_(other.closure_), enqueued_(other.enqueued_) {}
+
+WorkQueue::Storage& WorkQueue::Storage::operator=(Storage&& other) noexcept {
+  std::swap(closure_, other.closure_);
+  std::swap(enqueued_, other.enqueued_);
+  return *this;
+}
+
+EventEngine::Closure* WorkQueue::Storage::closure() { return closure_; }
+
+// ------ WorkQueue -----------------------------------------------------------
+
+// Returns whether the queue is empty
+bool WorkQueue::Empty() const {
+  return (most_recent_element_enqueue_timestamp_.load(
+              std::memory_order_relaxed) == kInvalidTimestamp &&
+          oldest_enqueued_timestamp_.load(std::memory_order_relaxed) ==
+              kInvalidTimestamp);
+}
+
+size_t WorkQueue::Size() {
+  grpc_core::MutexLock lock(&mu_);
+  return elements_.size() + (most_recent_element_enqueue_timestamp_.load(
+                                 std::memory_order_relaxed) == kInvalidTimestamp
+                                 ? 0
+                                 : 1);
+}
+
+grpc_core::Timestamp WorkQueue::OldestEnqueuedTimestamp() const {
+  int64_t front_of_queue_timestamp =
+      oldest_enqueued_timestamp_.load(std::memory_order_relaxed);
+  if (front_of_queue_timestamp != kInvalidTimestamp) {
+    return grpc_core::Timestamp::FromMillisecondsAfterProcessEpoch(
+        front_of_queue_timestamp);
+  }
+  int64_t most_recent_millis =
+      most_recent_element_enqueue_timestamp_.load(std::memory_order_relaxed);
+  if (most_recent_millis == kInvalidTimestamp) {
+    return grpc_core::Timestamp::InfPast();
+  }
+  return grpc_core::Timestamp::FromMillisecondsAfterProcessEpoch(
+      most_recent_millis);
+}
+
+absl::optional<EventEngine::Closure*> WorkQueue::PopFront()
+    ABSL_LOCKS_EXCLUDED(mu_) {
+  if (oldest_enqueued_timestamp_.load(std::memory_order_relaxed) !=
+      kInvalidTimestamp) {
+    absl::optional<EventEngine::Closure*> t = TryLockAndPop(/*front=*/true);
+    if (t.has_value()) return t;
+  }
+  if (most_recent_element_enqueue_timestamp_.load(std::memory_order_relaxed) !=
+      kInvalidTimestamp) {
+    return TryPopMostRecentElement();
+  }
+  return absl::nullopt;
+}
+
+absl::optional<EventEngine::Closure*> WorkQueue::PopBack() {
+  if (most_recent_element_enqueue_timestamp_.load(std::memory_order_relaxed) !=
+      kInvalidTimestamp) {
+    return TryPopMostRecentElement();
+  }
+  if (oldest_enqueued_timestamp_.load(std::memory_order_relaxed) !=
+      kInvalidTimestamp) {
+    absl::optional<EventEngine::Closure*> t = TryLockAndPop(/*front=*/false);
+    if (t.has_value()) return *t;
+  }
+  return absl::nullopt;
+}
+
+void WorkQueue::Add(EventEngine::Closure* closure) {
+  grpc_core::ExecCtx exec_ctx;
```

Comment:
is the exec_ctx to get current time?

Reply:
It is, that's its only purpose.

Reply:
as discussed, let's get https://github.com/grpc/grpc/pull/30815 in first and switch this to use that.